### PR TITLE
Remove ChannelReference and simplify Views

### PR DESCRIPTION
**Pull request by bleskes**
Currently we use ref counting to manage the life cycles of a translog file. This was done to allow the creation of view and snapshots, making sure that the underlying files are available. This commit takes a simpler route based on the observation that a snapshot doesn't need to have it's own life cycle but rather can lift on the lifecycle of it's parent (translog or view)[1].  Instead of ref counter, the channel is owned by the writer or reader pointing at it. Views are just a min generation indication and Snapshot get access to a channel they do not own (and thus are not closable).  This means ChannelReference and it's call backs on close are gone.

Also, I took the opportunity to clean up legacy translog readers we don't need in master.

[1] If code fails to adhere to this assumption it will get a channel already closed exception.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/translog/MultiSnapshot.java [hunk 1]
```diff
@@ -41,18 +36,7 @@
...
-            final int tops = translog.estimatedTotalOperations();
-            if (tops == TranslogReader.UNKNOWN_OP_COUNT) {
-                ops = TranslogReader.UNKNOWN_OP_COUNT;
-                break;
-            }
-            assert tops >= 0 : "tops must be positive but was: " + tops;
-            ops += tops;
-        }
-        estimatedTotalOperations = ops;
+        estimatedTotalOperations = Arrays.stream(translogs).mapToInt(Translog.Snapshot::estimatedTotalOperations).sum();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/MultiSnapshot.java]: given that all legacy is removed is this still an estimation?

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/MultiSnapshot.java]: Nope. Didn’t want to touch too much…  

Ps. Not sure when it was an estimation tbh.

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 2]
```diff
@@ -349,41 +341,49 @@ public long currentFileGeneration() {
...
+        return sizeInBytes(lastCommittedTranslogFileGeneration);
+    }
+
+    /**
+     * Returns the number of operations in the transaction files that aren't committed to lucene..
+     */
+    private int totalOperations(long minGeneration) {
+        try (ReleasableLock ignored = readLock.acquire()) {
+            ensureOpen();
+            return Stream.concat(readers.stream(), Stream.of(current))
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: nice!

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 3]
```diff
@@ -1348,24 +1223,61 @@ public void prepareCommit() throws IOException {
...
     public void commit() throws IOException {
-        ImmutableTranslogReader toClose = null;
         try (ReleasableLock lock = writeLock.acquire()) {
             ensureOpen();
-            if (currentCommittingTranslog == null) {
+            if (currentCommittingGeneration == NOT_SET_GENERATION) {
                 prepareCommit();
             }
+            assert currentCommittingGeneration != NOT_SET_GENERATION;
             lastCommittedTranslogFileGeneration = current.getGeneration(); // this is important - otherwise old files will not be cleaned up
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: can we assert that that generation has a reader in our readers list?

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 4]
```diff
@@ -1316,13 +1198,13 @@ public static void writeOperationNoSize(BufferedChecksumStreamOutput out, Transl
...
+            if (currentCommittingGeneration != NOT_SET_GENERATION) {
+                throw new IllegalStateException("already committing a translog with generation: " + currentCommittingGeneration);
             }
-            final TranslogWriter oldCurrent = current;
-            oldCurrent.ensureOpen();
-            oldCurrent.sync();
-            currentCommittingTranslog = current.immutableReader();
+            currentCommittingGeneration = current.generation;
+            current.sync();
+            ImmutableTranslogReader currentCommittingTranslog = current.closeIntoReader();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: maybe we can make sync a side-effect of closeIntoReader?

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 5]
```diff
@@ -1316,13 +1198,13 @@ public static void writeOperationNoSize(BufferedChecksumStreamOutput out, Transl
...
-            if (currentCommittingTranslog != null) {
-                throw new IllegalStateException("already committing a translog with generation: " + currentCommittingTranslog.getGeneration());
+            if (currentCommittingGeneration != NOT_SET_GENERATION) {
+                throw new IllegalStateException("already committing a translog with generation: " + currentCommittingGeneration);
             }
-            final TranslogWriter oldCurrent = current;
-            oldCurrent.ensureOpen();
-            oldCurrent.sync();
-            currentCommittingTranslog = current.immutableReader();
+            currentCommittingGeneration = current.generation;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: can generation be a getter?

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java (hunk 4 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: good catch, it already does:

From closeIntoReader:
```
sync();
ImmutableTranslogReader reader = new ImmutableTranslogReader(this.generation, channel, path, firstOperationOffset, getWrittenOffset(), operationCounter);
```

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 6]
```diff
@@ -1348,24 +1223,61 @@ public void prepareCommit() throws IOException {
...
+    }
+
+    void trimUnreferencedReaders() {
+        try (ReleasableLock ignored = writeLock.acquire()) {
+            if (closed.get()) {
+                // we're shutdown potentially on some tragic event - don't delete anything
+                return;
+            }
+            long minReferencedGen = outstandingViews.stream().mapToLong(View::minTranslogGeneration).min().orElse(Long.MAX_VALUE);
+            minReferencedGen = Math.min(lastCommittedTranslogFileGeneration, minReferencedGen);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: this loop is hard to follow. I think we can make this more elegant by using a sublist and make sure we process all of them.  we can also peak at the first element and get it's generation to calculate the range directly.

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 7]
```diff
@@ -1348,24 +1223,61 @@ public void prepareCommit() throws IOException {
...
+                IOUtils.closeWhileHandlingException(reader);
+                IOUtils.deleteFilesIgnoringExceptions(translogPath);
+                IOUtils.deleteFilesIgnoringExceptions(translogPath.resolveSibling(getCommitCheckpointFileName(reader.getGeneration())));
+            }
+        }
+    }
+
+    void closeFilesIfNoPendingViews() {
+        try (ReleasableLock ignored = writeLock.acquire()) {
+            if (closed.get() == false) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: can we do `if (closed.get() { // body }` to only have one return statement?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.Path;
+
+public class TranslogSnapshot extends TranslogReader implements Translog.Snapshot {
+    private final int totalOperations;
+    protected final long length;
+
+    private final ByteBuffer reusableBuffer;
+    long position;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java]: can these members be private please?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 9]
```diff
@@ -177,76 +177,43 @@ public long sizeInBytes() {
...
+        if (closed.compareAndSet(false, true)) {
+            boolean success = false;
+            try {
+                sync();
+                ImmutableTranslogReader reader = new ImmutableTranslogReader(this.generation, channel, path, firstOperationOffset, getWrittenOffset(), operationCounter);
+                success = true;
+                return reader;
+            } catch (Throwable e) {
+                closeWithTragicEvent(e);
+                throw new TranslogException(shardId, "exception while creating an immutable reader", e);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: I really wonder if we should just rethrow and get rid of TranslogException eventually

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 10]
```diff
@@ -177,76 +177,43 @@ public long sizeInBytes() {
...
-                } catch (Exception e) {
-                    throw new TranslogException(shardId, "exception while creating an immutable reader", e);
-                } finally {
-                    channelReference.decRef();
+    public synchronized ImmutableTranslogReader closeIntoReader() throws TranslogException, IOException {
+        if (closed.compareAndSet(false, true)) {
+            boolean success = false;
+            try {
+                sync();
+                ImmutableTranslogReader reader = new ImmutableTranslogReader(this.generation, channel, path, firstOperationOffset, getWrittenOffset(), operationCounter);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: s/this.generation/generation

**Comment by s1monw** [comment]
@bleskes thank you for working on this. I love the fact that we don't do ref counting anymore. I left a bunch of comments but it looks awesome. I also think @mikemccand should take a look at this.

File: core/src/main/java/org/elasticsearch/index/translog/MultiSnapshot.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/MultiSnapshot.java]: it was when we had to read until EOF to figure out how many :)

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/MultiSnapshot.java]: Yeah, but if memory serves it was either “I don’t know” or “I do know”. Anyway, long gone happily.

**Comment by bleskes** [comment]
Thanks @s1monw . Flight is delayed so I addressed all your comments :)

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 11]
```diff
@@ -122,8 +120,11 @@
     protected final ReleasableLock writeLock;
     private final Path location;
     private TranslogWriter current;
-    private volatile ImmutableTranslogReader currentCommittingTranslog;
-    private volatile long lastCommittedTranslogFileGeneration = -1; // -1 is safe as it will not cause an translog deletion.
+
+    private final static long NOT_SET_GENERATION = -1;
+
+    private volatile long currentCommittingGeneration = NOT_SET_GENERATION; // -1 is safe as it will not cause an translog deletion.
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: Move this comment up to the `NOT_SET_GENERATION` constant?  And "an translog" -> "a translog".

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 12]
```diff
@@ -122,8 +120,11 @@
...
+    private volatile long lastCommittedTranslogFileGeneration = NOT_SET_GENERATION; // -1 is safe as it will not cause an translog deletion.
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: Same here

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 13]
```diff
@@ -1348,24 +1223,61 @@ public void prepareCommit() throws IOException {
...
+        try (ReleasableLock ignored = writeLock.acquire()) {
+            if (closed.get() == false) {
+                // we are not closed yet.
+                return;
+            }
+            if (outstandingViews.isEmpty()) {
+                logger.trace("closing files. translog is closed and there are no pending views");
+                try {
+                    ArrayList<Closeable> toClose = new ArrayList<>(readers);
+                    toClose.add(current);
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: Can we move this `ArrayList` building outside of the `try`?

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 14]
```diff
@@ -1348,24 +1223,61 @@ public void prepareCommit() throws IOException {
...
+                    IOUtils.close(toClose);
+                } catch (IOException e) {
+                    logger.error("unexpected error while closing files", e);
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: Hmm why do we swallow this `IOException`...?

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 15]
```diff
@@ -112,7 +110,7 @@
 
     static final Pattern PARSE_STRICT_ID_PATTERN = Pattern.compile("^" + TRANSLOG_FILE_PREFIX + "(\\d+)(\\.tlog)$");
 
-    private final List<ImmutableTranslogReader> recoveredTranslogs;
+    private final List<ImmutableTranslogReader> readers = new ArrayList<>();
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: Can we add a comment here, that the readers are in sorted (by translog generation) order?

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 16]
```diff
@@ -112,7 +110,7 @@
...
     private volatile ScheduledFuture<?> syncScheduler;
     // this is a concurrent set and is not protected by any of the locks. The main reason
     // is that is being accessed by two separate classes (additions & reading are done by FsTranslog, remove by FsView when closed)
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: Fix comment: FsTranslog -> Translog, FsView -> View (a few other comments are also stale).

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 17]
```diff
@@ -600,151 +566,67 @@ public TranslogConfig getConfig() {
...
-        // last in this list is always FsTranslog.current
-        final List<TranslogReader> orderedTranslogs;
+        AtomicBoolean closed = new AtomicBoolean();
+        final long minGeneration;
         private final Callback<View> onClose;
 
-        View(List<TranslogReader> orderedTranslogs, Callback<View> onClose) {
-            // clone so we can safely mutate..
-            this.orderedTranslogs = new ArrayList<>(orderedTranslogs);
+        View(long minGeneration, Callback<View> onClose) {
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: Instead passing in a generic `onClose` callback here, can we just fold the `onViewClose.handle` logic directly into `View.close` below?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 18]
```diff
@@ -177,76 +177,43 @@ public long sizeInBytes() {
...
-        public int totalOperations() {
-            return TranslogWriter.this.totalOperations();
+    @Override
+    public synchronized Translog.Snapshot newSnapshot() {
+        try {
+            ensureOpen();
+            sync();
+            return super.newSnapshot();
+        } catch (IOException e) {
+            throw new TranslogException(shardId, "exception while a snapshot", e);
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: "while a snapshot" -> "while creating a snapshot"?

I think `ensureOpen` can be moved outside of the try?

Can we maybe separate `sync()` into its own try, with a different message ("exception while syncing for snapshot") on exception?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 19]
```diff
@@ -177,76 +177,36 @@ public long sizeInBytes() {
...
+    public synchronized ImmutableTranslogReader closeIntoReader() throws IOException {
         try {
-            final TranslogReader reader = new InnerReader(this.generation, firstOperationOffset, channelReference);
-            success = true;
-            return reader;
-        } finally {
-            if (!success) {
-                channelReference.decRef();
+            sync(); // sync before we close..
+            if (closed.compareAndSet(false, true)) {
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: Can we move this outside (after) the try?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java (hunk 18 above)
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: Should we call `closeWithTragicEvent(e)` here as well, if we hit `IOException` from `sync()`?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 20]
```diff
@@ -140,23 +145,18 @@ public void sync() throws IOException {
         if (syncNeeded()) {
             synchronized (this) {
                 ensureOpen(); // this call gives a better exception that the incRef if we are closed by a tragic event
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: Comment here is stale (no more incRef)?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Override
+    public final int totalOperations() {
+        return totalOperations;
+    }
+
+    @Override
+    public Translog.Operation next() throws IOException {
+        if (readOperations < totalOperations) {
+            assert readOperations < totalOperations : "readOperations must be less than totalOperations";
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java]: This assert is somewhat pointless?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 22]
```diff
@@ -140,23 +145,18 @@ public void sync() throws IOException {
...
-                channelReference.incRef();
+                final long offsetToSync;
+                final int opsCounter;
+                outputStream.flush();
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: Hmm, I think if this hits `Throwable` we should also `closeWithTragicEvent`, and we should javadoc this method that it declares tragedy on exceptions so callers don't need to?

**Comment by mikemccand** [comment]
I really like this simplification!  It's analogous to how in Lucene when we clone an `IndexInput` it's also not separately tracked but rather relies on its original `IndexInput` to remain open.

I left a bunch of small comments, many of which are likely pre-existing issues ... we can postpone/separate them from this simplification if it's a problem.

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java (hunk 14 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: It was the way of some interface before. Double checking now I think it's doable. Good catch.

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java (hunk 18 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: > Should we call closeWithTragicEvent(e) here as well, if we hit IOException from sync()?

I changed the sync to take care of this.

File: core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java (hunk 21 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java]: LOL

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java (hunk 22 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: haha. I changed this while working on other changes, before reaching the comment. Agreed :)  . I'll change the java docs.

**Comment by bleskes** [comment]
Thanks @mikemccand . I pushed another set of changes. Can you take another look?

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 23]
```diff
@@ -1348,24 +1213,53 @@ public void prepareCommit() throws IOException {
...
             }
-            toClose = this.currentCommittingTranslog;
-            this.currentCommittingTranslog = null;
-        } finally {
-            IOUtils.close(toClose);
+            long minReferencedGen = outstandingViews.stream().mapToLong(View::minTranslogGeneration).min().orElse(Long.MAX_VALUE);
+            minReferencedGen = Math.min(lastCommittedTranslogFileGeneration, minReferencedGen);
+            final long finalMinReferencedGen = minReferencedGen;
+            List<ImmutableTranslogReader> unreferenced = readers.stream().filter(r -> r.getGeneration() < finalMinReferencedGen).collect(Collectors.toList());
+            unreferenced.forEach(unreferencedReader -> {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: can we please not use `forEach` I think we should stick to for loops it just makes the code more readable and consistent

File: core/src/main/java/org/elasticsearch/index/translog/ImmutableTranslogReader.java [hunk 24]
```diff
@@ -19,50 +19,120 @@
...
-        if (channelReference.tryIncRef()) {
-            try {
-                ImmutableTranslogReader reader = newReader(generation, channelReference, firstOperationOffset, length, totalOperations);
-                channelReference.incRef(); // for the new object
-                return reader;
-            } finally {
-                channelReference.decRef();
+    /**
+     * Given a file, return a VersionedTranslogStream based on an
+     * optionally-existing header in the file. If the file does not exist, or
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/ImmutableTranslogReader.java]: This javadoc is stale?  The translog must exist, be non-zero length, have a proper header in 3.0?

File: core/src/main/java/org/elasticsearch/index/translog/ImmutableTranslogReader.java [hunk 25]
```diff
@@ -88,4 +158,20 @@ public Checkpoint getInfo() {
...
+        }
+    }
+
+    protected final boolean isClosed() {
+        return closed.get();
+    }
+
+    protected void ensureOpen() {
+        if (isClosed()) {
+            throw new AlreadyClosedException("translog [" + getGeneration() + "] is already closed");
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/ImmutableTranslogReader.java]: Can we change `getGeneration()` to `toString()` instead so the exception also tells us which file was problematic?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogReader.java [hunk 26]
```diff
@@ -19,45 +19,30 @@
...
-import java.util.concurrent.atomic.AtomicBoolean;
 
 /**
  * A base class for all classes that allows reading ops from translog files
  */
-public abstract class TranslogReader implements Closeable, Comparable<TranslogReader> {
-    public static final int UNKNOWN_OP_COUNT = -1;
-    private static final byte LUCENE_CODEC_HEADER_BYTE = 0x3f;
-    private static final byte UNVERSIONED_TRANSLOG_HEADER_BYTE = 0x00;
+public abstract class TranslogReader implements Comparable<TranslogReader> {
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogReader.java]: Can we rename this class to `BaseTranslogReader` and rename `ImmutableTranslogReader` to `TranslogReader`?

File: core/src/main/java/org/elasticsearch/index/translog/ImmutableTranslogReader.java [hunk 27]
```diff
@@ -19,50 +19,120 @@
...
+                        BytesRef ref = new BytesRef(len);
+                        ref.length = len;
+                        headerStream.read(ref.bytes, ref.offset, ref.length);
+                        BytesRef uuidBytes = new BytesRef(translogUUID);
+                        if (uuidBytes.bytesEquals(ref) == false) {
+                            throw new TranslogCorruptedException("expected shard UUID [" + uuidBytes + "] but got: [" + ref + "] this translog file belongs to a different translog");
+                        }
+                        return new ImmutableTranslogReader(checkpoint.generation, channel, path, ref.length + CodecUtil.headerLength(TranslogWriter.TRANSLOG_CODEC) + RamUsageEstimator.NUM_BYTES_INT, checkpoint.offset, checkpoint.numOps);
+                    default:
+                        throw new TranslogCorruptedException("No known translog stream version: " + version + " path:" + path);
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/ImmutableTranslogReader.java]: It's great that this exception message includes the path ... can we fix the others to do so as well?

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 28]
```diff
@@ -502,26 +476,9 @@ private static Snapshot createSnapshot(TranslogReader... translogs) {
...
-                // (we hold a read lock) and the view will be notified of any future one
-                outstandingViews.add(view);
-                translogs.clear();
-                return view;
-            } finally {
-                // close if anything happend and we didn't reach the clear
-                IOUtils.closeWhileHandlingException(translogs);
-            }
-        }
+        View view = new View(lastCommittedTranslogFileGeneration);
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: Shouldn't we have an `ensureOpen` here?

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 29]
```diff
@@ -1348,24 +1213,53 @@ public void prepareCommit() throws IOException {
...
+            minReferencedGen = Math.min(lastCommittedTranslogFileGeneration, minReferencedGen);
+            final long finalMinReferencedGen = minReferencedGen;
+            List<ImmutableTranslogReader> unreferenced = readers.stream().filter(r -> r.getGeneration() < finalMinReferencedGen).collect(Collectors.toList());
+            unreferenced.forEach(unreferencedReader -> {
+                Path translogPath = unreferencedReader.path();
+                logger.trace("delete translog file - not referenced and not current anymore {}", translogPath);
+                IOUtils.closeWhileHandlingException(unreferencedReader);
+                IOUtils.deleteFilesIgnoringExceptions(translogPath);
+                IOUtils.deleteFilesIgnoringExceptions(translogPath.resolveSibling(getCommitCheckpointFileName(unreferencedReader.getGeneration())));
+            });
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: Hmm why are we ignoring exceptions here?  You can consolidate those two `deleteFilesIgnoringExceptions` into one call and it will do the right thing if either path hit an exc while being deleted...

File: core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java [hunk 30]
```diff
@@ -0,0 +1,104 @@
...
+    public long sizeInBytes() {
+        return length;
+    }
+
+    /**
+     * reads an operation at the given position into the given buffer.
+     */
+    protected void readBytes(ByteBuffer buffer, long position) throws IOException {
+        if (position >= length) {
+            throw new EOFException("read requested past EOF. pos [" + position + "] end: [" + length + "]");
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java]: Can we get the translog path into this exc?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java [hunk 31]
```diff
@@ -0,0 +1,104 @@
...
+        }
+        if (position < getFirstOperationOffset()) {
+            throw new IOException("read requested before position of first ops. pos [" + position + "] first op on: [" + getFirstOperationOffset() + "]");
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java]: Also in this one?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 32]
```diff
@@ -177,76 +182,36 @@ public long sizeInBytes() {
...
-                    throw new TranslogException(shardId, "exception while creating an immutable reader", e);
-                } finally {
-                    channelReference.decRef();
-                }
-            }
+        if (closed.compareAndSet(false, true)) {
+            return new ImmutableTranslogReader(generation, channel, path, firstOperationOffset, getWrittenOffset(), operationCounter);
         } else {
-            throw new TranslogException(shardId, "can't increment channel [" + channelReference + "] ref count");
+            throw new AlreadyClosedException("translog [" + getGeneration() + "] is already closed", tragedy);
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: Can we get the translog path here too?

**Comment by mikemccand** [comment]
Thanks @bleskes, I left a few minor comments, most of them pre-existing issues so feel free to defer to separate issue, but otherwise LGTM: this is a nice simplification.

File: core/src/main/java/org/elasticsearch/index/translog/ImmutableTranslogReader.java (hunk 25 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/ImmutableTranslogReader.java]: good idea

File: core/src/main/java/org/elasticsearch/index/translog/TranslogReader.java (hunk 26 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogReader.java]: Will do so, but I'll merge master first to reduce conflicts..

File: core/src/main/java/org/elasticsearch/index/translog/ImmutableTranslogReader.java (hunk 27 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/ImmutableTranslogReader.java]: will do

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java (hunk 28 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: added. It requires acquiring a read lock  to be effecitve.

File: core/src/main/java/org/elasticsearch/index/translog/Translog.java (hunk 29 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/Translog.java]: I merged the two calls. This is old code. I rather not open the can of worms of not ignoring the exceptions. I agree with you though :)

File: core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java (hunk 30 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java]: done

File: core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java (hunk 31 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogSnapshot.java]: ++

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java (hunk 32 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: added

**Comment by bleskes** [comment]
@mikemccand @s1monw thanks. I pushed some more commits addressing your feedback

**Comment by mikemccand** [comment]
LGTM, thanks @bleskes, this is a great change!

**Comment by bleskes** [comment]
Pushed to master. Thanks @mikemccand and @s1monw , these are tricky reviews :)
